### PR TITLE
Fix multidex

### PR DESCRIPTION
**Pull request by Kright**
new dsl for multidex. 
new feature - including in maindex only necessary classes

### CODE REVIEW

File: examples/HelloScalaApp/hellolibrary/build.gradle [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+			minifyEnabled false
+			proguardFiles getDefaultProguardFile('proguard-android.txt'), 'proguard-rules.pro'
+		}
+	}
+}
+
+androidScala {
+	scalaVersion '2.11.8'
+	zincVersion '0.3.12'
+	addScalaLibrary true
```
Comment (faqqe) [inline examples/HelloScalaApp/hellolibrary/build.gradle]: have we not agreed that this doesn't make sense? what is the scenario when you dont want to add scala library? even if this scenario existed, library should be added by default and excluding it should be explicit

File: plugin/src/main/groovy/com/github/kright/androidscalasupport/AndroidScalaSupport.groovy [hunk 2]
```diff
@@ -59,29 +88,71 @@ class AndroidScalaSupport implements Plugin<Project> {
...
-	private def setScalaCompileTaskAdding() {
-		project.afterEvaluate {
-			androidExtension.applicationVariants.all { addScalaCompile(it) }
-			androidExtension.testVariants.all { addScalaCompile(it) }
-			androidExtension.unitTestVariants.all { addScalaCompile(it) }
+	/**
+	 * creates MainDexModifier and sets callbacks
+	 */
+	private setMainDexModification() {
+		if (isLibrary) return
```
Comment (faqqe) [inline plugin/src/main/groovy/com/github/kright/androidscalasupport/AndroidScalaSupport.groovy]: code style: usually single-line `if` is only used if it is an `if-else` expression
`val x = if (a) 0 else 1`
single-line `if`-statement is confusing

File: plugin/src/main/groovy/com/github/kright/androidscalasupport/AndroidScalaSupport.groovy [hunk 3]
```diff
@@ -59,29 +88,71 @@ class AndroidScalaSupport implements Plugin<Project> {
...
+		androidExtension.applicationVariants.all { variant ->
+			def name = variant.name.capitalize()
+
+			project.tasks.getByName("process${name}Manifest").doLast { manifestProcessorTask ->
+				if (extension.multiDex.mainDexModifier == null)
+					return
+
+				File manifest = manifestProcessorTask.manifestOutputFile
+
+				project.tasks.getByName("transformClassesWithDexFor$name").doFirst { transformTask ->
```
Comment (faqqe) [inline plugin/src/main/groovy/com/github/kright/androidscalasupport/AndroidScalaSupport.groovy]: whats up with weird formatting? is it you or github displays incorrectly?

File: plugin/src/main/groovy/com/github/kright/androidscalasupport/AndroidScalaSupport.groovy [hunk 4]
```diff
@@ -111,7 +186,14 @@ class AndroidScalaSupport implements Plugin<Project> {
...
 
-	private def configuration(name, dependency) {
+	/**
+	 * creates new configuration or return the existing
+	 *
+	 * @param name of the configuration
+	 * @param dependency of the configuration
+	 * @return project configuration
+	 */
+	private configuration(name, dependency) {
```
Comment (faqqe) [inline plugin/src/main/groovy/com/github/kright/androidscalasupport/AndroidScalaSupport.groovy]: add types

File: plugin/src/main/groovy/com/github/kright/androidscalasupport/MainDexModifier.groovy [hunk 5]
```diff
@@ -0,0 +1,65 @@
+package com.github.kright.androidscalasupport
+
+/**
+ * overwrites given maindexlist.txt
+ *
+ * Created by lgor on 01.03.2017.
+ */
+class MainDexModifier {
+
+	private final static multiDexClasses = [
```
Comment (faqqe) [inline plugin/src/main/groovy/com/github/kright/androidscalasupport/MainDexModifier.groovy:10]: you could prepend the values in the array with 'android/support/multiDex/' part by using `map`

File: plugin/src/main/groovy/com/github/kright/androidscalasupport/MainDexModifier.groovy [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+
+	MainDexModifier(AndroidScalaSupport androidScalaSupport) {
+		assert androidScalaSupport != null
+		plugin = androidScalaSupport
+	}
+
+	private List classesFromManifest(File manifestFile) {
+		assert manifestFile != null
+		assert manifestFile.exists()
+		assert manifestFile.isFile()
```
Comment (faqqe) [inline plugin/src/main/groovy/com/github/kright/androidscalasupport/MainDexModifier.groovy]: why

File: plugin/src/main/groovy/com/github/kright/androidscalasupport/MainDexOverwrite.groovy [hunk 7]
```diff
@@ -0,0 +1,24 @@
+package com.github.kright.androidscalasupport
+
+/**
+ * Created by lgor on 01.03.2017.
+ */
```
Comment (faqqe) [inline plugin/src/main/groovy/com/github/kright/androidscalasupport/MainDexOverwrite.groovy]: `MainDexOverwrite` is a strange name, consider `MainDexRules` or `MainDexOptions`

File: plugin/src/main/groovy/com/github/kright/androidscalasupport/MultidexExtension.groovy [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+	private boolean enabled = false
+
+	MultidexExtension(Project project) {
+		this.project = project
+	}
+
+	/**
+	 * enables multiDex
+	 */
+	def enabled(boolean value) {
```
Comment (faqqe) [inline plugin/src/main/groovy/com/github/kright/androidscalasupport/MultidexExtension.groovy]: `enabled` is a name for a property, not a method
you mean `enable`?

File: plugin/src/main/groovy/com/github/kright/androidscalasupport/MultidexExtension.groovy [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+import org.gradle.api.Project
+
+/**
+ * Created by lgor on 01.03.2017.
+ */
+class MultidexExtension {
+
+	private final Project project
+
+	private MainDexOverwrite mainDexModifier = null
```
Comment (faqqe) [inline plugin/src/main/groovy/com/github/kright/androidscalasupport/MultidexExtension.groovy]: usually you want the class and the method named the same if theres no particular reason not to

File: plugin/src/main/groovy/com/github/kright/androidscalasupport/MultidexExtension.groovy (hunk 8 above)
Comment (Kright) [reply plugin/src/main/groovy/com/github/kright/androidscalasupport/MultidexExtension.groovy]: when user writes:
```
multiDex.enabled true 
```
He actually calls this method

File: README.md [hunk 10]
```diff
@@ -19,17 +18,48 @@ buildscript {
...
+
 ```groovy
-// needs to be applied after android plugin
+// has to be applied after andoid plugin
 apply plugin: 'android-scala-support'
+```
+
+In common cases scalaVersion and multidex are required:
+```
+adnroidScala {
```
Comment (faqqe) [inline README.md]: type (android)

File: plugin/src/main/groovy/com/github/kright/androidscalasupport/AndroidScalaExtension.groovy [hunk 11]
```diff
@@ -1,65 +1,122 @@
...
-				compile 'com.android.support:multidex:1.0.1'
+		/**
+		 * @return MainDexOverwriteRule. may be null
+		 */
+		MainDexOverwriteRule getMainDexOverwriteRule() {
+			return dexOverwriteRule
+		}
+
+
+		class MainDexOverwriteRule {
```
Comment (faqqe) [inline plugin/src/main/groovy/com/github/kright/androidscalasupport/AndroidScalaExtension.groovy]: plural

File: plugin/src/main/groovy/com/github/kright/androidscalasupport/AndroidScalaSupport.groovy [hunk 12]
```diff
@@ -59,29 +81,73 @@ class AndroidScalaSupport implements Plugin<Project> {
...
 
-	private def setScalaCompileTaskAdding() {
-		project.afterEvaluate {
-			androidExtension.applicationVariants.all { addScalaCompile(it) }
-			androidExtension.testVariants.all { addScalaCompile(it) }
-			androidExtension.unitTestVariants.all { addScalaCompile(it) }
+	/**
+	 * creates MainDexModifier and sets callbacks
+	 */
+	private setMainDexModification() {
```
Comment (faqqe) [inline plugin/src/main/groovy/com/github/kright/androidscalasupport/AndroidScalaSupport.groovy]: not a setter

File: plugin/src/main/groovy/com/github/kright/androidscalasupport/AndroidScalaSupport.groovy [hunk 13]
```diff
@@ -59,29 +81,73 @@ class AndroidScalaSupport implements Plugin<Project> {
...
+				}
+			}
 		}
 	}
 
-	private def addScalaCompile(variant) {
+	/**
+	 * creates ScalaCompile tasks for each build variant
+	 */
+	private setScalaCompileTaskAdding() {
```
Comment (faqqe) [inline plugin/src/main/groovy/com/github/kright/androidscalasupport/AndroidScalaSupport.groovy]: installScalaCompile

**Review by faqqe — CHANGES_REQUESTED** [review]
fix typos

**Review by faqqe — APPROVED** [review]

**Pull request merged**